packages/amplify-category-auth/provider-utils/awscloudformation/triggers/PostConfirmation/function-template-dir/trigger-index.js

```diff
@@ -1,14 +1,51 @@
-/*
-  this file will loop through all js modules which are uploaded to the lambda resource,
-  provided that the file names (without extension) are included in the "MODULES" env variable.
-  "MODULES" is a comma-delimmited string.
-*/
+/**
+ * @fileoverview
+ *
+ * This CloudFormation Trigger creates a Lambda function which waits for all
+ * other specified Lambdas (modules) to resolve, which should be located in the
+ * same directory as this file (./).
+ */
+
+/**
+ * The names of modules to load are stored as a comma-delimited string in the
+ * `MODULES` env variable.
+ */
 const moduleNames = process.env.MODULES.split(',');
+/**
+ * The array of imported modules.
+ */
 const modules = moduleNames.map(name => require(`./${name}`));
 
-exports.handler = (event, context, callback) => {
-  for (let i = 0; i < modules.length; i += 1) {
-    const { handler } = modules[i];
-    handler(event, context, callback);
-  }
+/**
+ * This handler is itself a Lambda function which iterates over all of the given
+ * modules and awaits them.
+ *
+ * @see https://docs.aws.amazon.com/lambda/latest/dg/nodejs-handler.html
+ *
+ * @param {object} event
+ *
+ * The event that triggered this Lambda.
+ *
+ * @param {object} context 
+ *
+ * The context for this Lambda. See:
+ * https://docs.aws.amazon.com/lambda/latest/dg/nodejs-context.html
+
+ * @param {function} callback 
+ *
+ * A deprecated way to send a response synchronously. See:
+ * https://docs.aws.amazon.com/lambda/latest/dg/nodejs-handler.html#nodejs-handler-sync
+ *
+ * @returns
+ *
+ * The CloudFormation Trigger event.
+ */
+exports.handler = async (event, context, callback) => {
+  /**
+   * Instead of naively iterating over all handlers, run them concurrently with
+   * `await Promise.all(...)`. This would otherwise just be determined by the
+   * order of names in the `MODULES` var.
+   */
+  await Promise.all(modules.map(module => module.handler(event, context, callback)));
+  return event;
```

Comment:
Is there a reason for returning the event? Seems like just awaiting the promise and returning void would be better?

Reply:
So, I forgot to really elaborate on this, I have been absolutely slammed for time, but I am pretty sure that is the real error here, was the generated top-level Lambda not returning anything.

I found that even with async/await implemented, I would get the JSON invalid error. After a few minutes, really getting a grasp on what's going on here, that top-level lambda is called by CloudFormation as I understand it, and that Lambda loads all the modules for that step (here, `PostConfirmation`). If that top-level Lambda passes CF back an event of `undefined`, it seems to throw the Invalid JSON error.

Basically, the theory is that even if I hadn't patched this with async/await, I could have added this return statement and patched that error, and the async stuff would have been running on its own out of order (not ideal) but the error would have disappeared.

We should look into this more IMO, and if I'm right, it implies that we should *really* go over all the Amplify templates, move them to async/await, and replace the `index-trigger.js` with this one right here.

Is my theory making sense? CloudFormation wants an object back from that trigger. If there's another layer in the stack after one of these Lambdas that doesn't return, that stack will get passed an invalid event (nothing) and throw "invalid JSON".

I would look into this more but as I said I am really out of time. Please let me know what you think and if you can confirm or unearth more here.

I tried to help someone in this thread with that trick, his trigger was totally synchronous so it seems consistent with these triggers not returning anything causing problems with CloudFormation, but it didn't solve his problem:

https://github.com/aws-amplify/amplify-cli/issues/4341#issuecomment-828140643

Reply:
Also, if we confirm this, I will add a mention of it to the `@fileoverview`.

Reply:
Consistent with this, the generated Lambda for Verification Link (which was the issue that fellow had in that thread) also 400s out despite mostly working. Case for broken `index-trigger.js` strong, basically every CloudFormation stack created by this CLI with more than one layer has been broken for some indeterminate amount of time it would seem.

![image](https://user-images.githubusercontent.com/1657236/116651496-f51d2700-a948-11eb-8ed1-93475f5b2226.png)

Reply:
Confirmed, that updating the new `trigger-index.js` (with async/await support + returning event) *for each trigger in the stack* and updating the submodules to use modern syntax (`async` + return or throw), fixes this issue and also #4341.

Reply:
See #7219.